### PR TITLE
Update RequiresPreviewFeatures attribute

### LINKED ISSUE #56498: Add description to RequiresPreviewFeaturesAttribute
It would be nice if there was a way to provide a description with `RequiresPreviewFeaturesAttribute`, like you can with `ObsoleteAttribute`. The description would show up in analyzers/build warnings.

A description would provide more information about why a feature is in preview and what a user should keep in mind if they choose to use the feature.

For example, HTTP/3 in .NET 6 is preview. Kind of detail we want to pass on when someone sees this warning:
* What scenarios HTTP/3 does and doesn't support in .NET 6.
* How to mitigate an app if HTTP/3 is turned on and problems are found.

### API Proposal

```C#
namespace System.Runtime.Versioning
{
    public sealed class RequiresPreviewFeaturesAttribute : Attribute
    {
        // Existing
        // public RequiresPreviewFeaturesAttribute();
        public RequiresPreviewFeaturesAttribute(string message);
        public string Message { get; }
        public string Url { get; set; }
    }
}
```

Note: The URL property allows us to surface the URL directly via the Roslyn diagnostic, so that people can click the diagnostic ID in the error list, rather than copy & pasting the URL.

### API Usage

```diff
 namespace Microsoft.AspNetCore.Server.Kestrel.Core
 {
     [Flags]
     public enum HttpProtocols
     {
         Http1 = 0x1,
         Http2 = 0x2,
         Http1AndHttp2 = Http1 | Http2,
+        [RequiresPreviewFeatures("Kestrel HTTP3 support for .NET 6 is in preview", Url = "https://aka.ms/kestrel-http3")]
         Http3 = 0x4,
+        [RequiresPreviewFeatures("Kestrel HTTP3 support for .NET 6 is in preview", Url = "https://aka.ms/kestrel-http3")]
         Http1AndHttp2AndHttp3 = Http1 | Http2 | Http3
     }
 }
```

### PR DESCRIPTION
**Pull request by pgovind**
Updates RequiresPreviewFeatures attribute with an optional string message and URL

Fixes https://github.com/dotnet/runtime/issues/56498

cc @jeffhandley @terrajobst

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs [hunk 1]
```diff
@@ -10,7 +10,38 @@ public class RequiresPreviewFeaturesAttributeTests
...
+        {
+            var attribute = new RequiresPreviewFeaturesAttribute(message) {};
+            Assert.Equal(message, attribute.Message);
+            Assert.Null(attribute.Url);
+        }
+
+        [Theory]
+        [InlineData(null, "")]
+        [InlineData("", null)]
+        [InlineData("message", "https://aka.ms/obsolete/{0}")]
```
Comment (pgovind) [inline src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs:39]: Fix input string

File: src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs [hunk 2]
```diff
@@ -10,7 +10,38 @@ public class RequiresPreviewFeaturesAttributeTests
         [Fact]
         public void RequiresPreviewFeaturesAttributeTest()
         {
-            new RequiresPreviewFeaturesAttribute();
+            var attribute = new RequiresPreviewFeaturesAttribute();
+            Assert.NotNull(attribute);
```
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs:14]: This is unnecessary :)

File: src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs [hunk 3]
```diff
@@ -10,7 +10,38 @@ public class RequiresPreviewFeaturesAttributeTests
...
+            Assert.Null(attribute.Url);
+        }
+
+        [Theory]
+        [InlineData(null)]
+        [InlineData("")]
+        [InlineData("message")]
+        public void Ctor_String_Message(string message)
+        {
+            var attribute = new RequiresPreviewFeaturesAttribute(message) {};
```
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs:31]:
```suggestion
            var attribute = new RequiresPreviewFeaturesAttribute(message);
```

File: src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs [hunk 4]
```diff
@@ -10,7 +10,38 @@ public class RequiresPreviewFeaturesAttributeTests
...
+            Assert.Equal(message, attribute.Message);
```
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs:32]:
```suggestion
            Assert.Same(message, attribute.Message);
```

File: src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs [hunk 5]
```diff
@@ -10,7 +10,38 @@ public class RequiresPreviewFeaturesAttributeTests
...
+
+        [Theory]
+        [InlineData(null, "")]
+        [InlineData("", null)]
+        [InlineData("message", "https://aka.ms/preview-features/")]
+        public void Ctor_String_Url(string message, string url)
+        {
+            var attribute = new RequiresPreviewFeaturesAttribute(message) { Url = url };
+            Assert.Equal(message, attribute.Message);
+            Assert.Equal(url, attribute.Url);
```
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/Runtime/Versioning/RequiresPreviewFeaturesAttributeTests.cs:44]:
```suggestion
            Assert.Same(message, attribute.Message);
            Assert.Same(url, attribute.Url);
```

**Review by stephentoub — APPROVED** [review]
Thanks.

**Review by jeffhandley — APPROVED** [review]

**Comment by pgovind** [comment]
Unrelated build failures. Merging. Thanks for the reviews!!

**Comment by JamesNK** [comment]
Hi, can the ASP.NET Core repo be updated to use the message immediately? Will it just work once the runtime changes flow into aspnetcore and aspnetcore includes a message, or does it also require other changes in VS or the compiler for the message to show up when building?

**Pull request merged**